### ISSUE TITLE
`handleResponse` on the adapter is never called when using fetch and a response with no body

### ISSUE THREAD

**Issue by kategengler** [emberjs/data#6851]
When using `fetch`, a response with a status and an empty body never hits `handleResponse` on the application adapter. 

If I turn `jquery-integration` back on, such that `useFetch` is true, then handleResponse is hit, for the same response code and empty body.

Jumping through the debugger, I can follow it down to `response.text()` in `determineBodyPromise`, but it never makes it to the `then` on that promise. It seems like there may be a missing `catch` in the chain of promises here https://github.com/emberjs/data/blob/master/packages/adapter/addon/rest.js#L1002  

### Versions

Run the following command and paste the output below: `yarn list ember-source && yarn list ember-cli && yarn list --pattern ember-data`.

```cli
└─ ember-source@3.14.2

└─ ember-cli@3.14.0

├─ @ember-data/-build-infra@3.14.0
├─ @ember-data/adapter@3.14.0
├─ @ember-data/canary-features@3.14.0
├─ @ember-data/model@3.14.0
├─ @ember-data/rfc395-data@0.0.4
├─ @ember-data/serializer@3.14.0
├─ @ember-data/store@3.14.0
└─ ember-data@3.14.0
```

**Comment by runspired** [comment]
I think there's three changes to make and add test coverage for here:

- we need to catch and handle if determineBodyPromise errors
- we need to make sure if `fetch` itself rejects we still catch that and process correctly
- we should fix determineBodyPromise's ability to handle empty payloads

@Gaurav0 mind taking a look?

**Comment by andreyfel** [comment]
I think I encountered a similar issue. When my server returns 401 with an empty body I'm getting "SyntaxError: Unexpected end of JSON input" in the application route error action instead of the 401 error.
It started happening in Ember Data 3.14. It worked as expected in ED 3.13.

A change in this file caused the issue:
packages/adapter/addon/-private/utils/determine-body-promise.ts
https://github.com/emberjs/data/pull/6293

There is a new `throw` that breaks the chain. Interesting that the throw happens only in `if (DEBUG)` check. So, in the production build the behavior would be different.

**Comment by kategengler** [comment]
@andreyfel I think that you may be seeing the same bug, but I don't think the PR you linked to is the cause: In my testing, it never made it into the `then` on `response.text()` in that method, so it would never reach the code that was added.

**Comment by andreyfel** [comment]
@kategengler In my case I've set a breakpoint on the code I was talking about and I could see that it was reached.

**Comment by kategengler** [comment]
@andreyfel It may be a different bug then, as I did the same for just inside the `then` on `response.text()` and the breakpoint was never hit.

**Comment by andreyfel** [comment]
@kategengler can you check if your issue can be reproduced with ED 3.13?

**Comment by runspired** [comment]
This should be resolved by #6871

**Comment by runspired** [comment]
Fix can be tested with `3.16.0-alpha.3`
